### ISSUE TITLE
Maneuvers API

### ISSUE THREAD

**Issue by VitaminPSG** [mapbox/mapbox-navigation-android#4576]
In the process of developing the functionality of the maneuvers, some of the requirements could not be implemented using the current API. Below is a list of requirements that I would like to add to the current SDK

- As a user, I can see the location of the maneuver on the map
> This requires adding a `Point` to the maneuver model.
>
>![device-2021-07-06-134311](https://user-images.githubusercontent.com/5839897/124587654-85ec1380-de60-11eb-9b45-84af5a959f49.gif)

- In the list of maneuvers, I can see my destination
> In the current implementation, there is no destination card, in most cases it is a common description that your destination is on the right or left.
>
>![device-2021-07-06-135508](https://user-images.githubusercontent.com/5839897/124588746-d31cb500-de61-11eb-8637-0d134b4e875b.png)

- I can apply styles for inactive turn lines
> In the current version, transparency is applied for inactive turn line, where we want to change its color to our brand design
>
> ![device-2021-07-06-140114](https://user-images.githubusercontent.com/5839897/124589427-a917c280-de62-11eb-8adf-c687c8f4b205.png)

### Bugs:
 - `MapboxLaneGuidanceAdapter` should arrange turn lines across the entire width of the view
 - Check all possible modifications for `Lane` to exclude empty cells. See image above
 - `UpcomingManeuverAdapter` should use same formatter as `MapboxManeuverApi`
 - `MapboxManeuverApi` shouldn't return duplicates for maneuvers
>
><img width="1280" alt="Screen Shot 2021-06-29 at 11 35 28" src="https://user-images.githubusercontent.com/5839897/124590903-550ddd80-de64-11eb-9589-b1d4ae969d88.png">

 - ApiCalls shouldn't be suspend functions 

affected: https://github.com/mapbox/gm-navigation-app/pull/2205
cc @LukasPaczos @abhishek1508

**Comment by abhishek1508** [comment]
@VitaminPSG 

Thank you for submitting the feedback as it's very helpful and will definitely be handy for us to make our API and Views better.

> As a user, I can see the location of the maneuver on the map

This makes a lot of sense to me and I think we should do it asap as it also meets our vision of next iteration of maneuvers.

> In the list of maneuvers, I can see my destination

I don't think that the SDK should support this use case. It is very user centric and different customers can have very different requirements. It is very simple to add a destination as the last item in the upcoming maneuver list and can be very easily added in the view after retrieving the list from the API.

> I can apply styles for inactive turn lines

You can change the colors for inactive turn lane indicators. To do so you would have to do the following:
```
<style name="MyCustomStyleTurnIconManeuver">
    <item name="maneuverTurnIconColor">your_active_color</item>
    <item name="maneuverTurnIconShadowColor">your_inactive_color</item>
</style>
```
Apply this style as follows:
```
mapboxManeuverView.updateTurnIconStyle(R.style. MyCustomStyleTurnIconManeuver)
```

However, I agree that the SDK will apply alpha to the inactive color no matter whatever color you provide. This is something, I feel SDK should remove and allow developers to specify any alpha to the color itself if needed. For ex: instead of #969696, you could provide a color value #80969696

> MapboxLaneGuidanceAdapter should arrange turn lines across the entire width of the view

This is purely a design consideration and I would leave it up to the design engineers to choose the best design based on user research. cc @d-prukop 

> Check all possible modifications for Lane to exclude empty cells. See image above

There are 1000's of possible lane combinations which is very difficult for the SDK to know beforehand. If you are seeing missing empty cells that means the SDK does not have access to the proper asset for that specific lane combination. We are well aware of this issue.

> UpcomingManeuverAdapter should use same formatter as MapboxManeuverApi

This has already been fixed in our latest iteration.

> MapboxManeuverApi shouldn't return duplicates for maneuvers

We are aware of this bug and are working on it.

> ApiCalls shouldn't be suspend functions

This has already been fixed in our latest iteration.

cc @LukasPaczos

**Comment by VitaminPSG** [comment]
@abhishek1508 Is it possible to show maneuvers for all legs not for the current one?

**Comment by LukasPaczos** [comment]
> Is it possible to show maneuvers for all legs not for the current one?

@VitaminPSG you can do:
```
maneuverApi.getManeuvers(routeProgress) // current leg
maneuverApi.getManeuvers(routeProgress.route, routeProgress.currentLegProgress.legIndex + 1) // upcoming leg
etc...
```

**Comment by abhishek1508** [comment]
> @abhishek1508 Is it possible to show maneuvers for all legs not for the current one?

@VitaminPSG While refactoring the API we thought about this use case, but disregarded thinking that it doesn't make sense to show maneuvers for all the route legs at once when you are navigating on just one of them. As @LukasPaczos already shared an example how you could still do it, I would love to understand the UI requirements that would need you to fetch maneuvers for all route legs.

**Comment by LukasPaczos** [comment]
@abhishek1508 would you be able to branch off anything left to do here into separate tickets and close this one?

**Comment by abhishek1508** [comment]
> In the process of developing the functionality of the maneuvers, some of the requirements could not be implemented using the current API. Below is a list of requirements that I would like to add to the current SDK
> 
> * As a user, I can see the location of the maneuver on the map
> 
> > This requires adding a `Point` to the maneuver model.
> > ![device-2021-07-06-134311](https://user-images.githubusercontent.com/5839897/124587654-85ec1380-de60-11eb-9b45-84af5a959f49.gif)

**Available in v2.0.0**
```
class Maneuver internal constructor(
    val primary: PrimaryManeuver,
    val stepDistance: StepDistance,
    val secondary: SecondaryManeuver? = null,
    val sub: SubManeuver? = null,
    val laneGuidance: Lane? = null,
    val maneuverPoint: Point
)
```
> 
> * In the list of maneuvers, I can see my destination
> 
> > In the current implementation, there is no destination card, in most cases it is a common description that your destination is on the right or left.
> > ![device-2021-07-06-135508](https://user-images.githubusercontent.com/5839897/124588746-d31cb500-de61-11eb-8637-0d134b4e875b.png)

**Needs design input**. We probably don't want to do it because this behavior is very specific to this use case but might not be required by other customers. https://github.com/mapbox/mapbox-navigation-android/issues/5063
> 
> * I can apply styles for inactive turn lines
> 
> > In the current version, transparency is applied for inactive turn line, where we want to change its color to our brand design
> > ![device-2021-07-06-140114](https://user-images.githubusercontent.com/5839897/124589427-a917c280-de62-11eb-8adf-c687c8f4b205.png)

**Available in v2.0.0**
> 
> ### Bugs:
> * `MapboxLaneGuidanceAdapter` should arrange turn lines across the entire width of the view

https://github.com/mapbox/mapbox-navigation-android/issues/5064
> * Check all possible modifications for `Lane` to exclude empty cells. See image above

@VitaminPSG Could you please confirm if this still happens? If yes, could you please share a sample origin/destination pair?

> * `UpcomingManeuverAdapter` should use same formatter as `MapboxManeuverApi`

**Available in v2.0.0**
> * `MapboxManeuverApi` shouldn't return duplicates for maneuvers
> 
> > <img alt="Screen Shot 2021-06-29 at 11 35 28" width="1280" src="https://user-images.githubusercontent.com/5839897/124590903-550ddd80-de64-11eb-9589-b1d4ae969d88.png">

**Available in v2.0.0**

> * ApiCalls shouldn't be suspend functions

**Available in v2.0.0**
> affected: [mapbox/gm-navigation-app#2205](https://github.com/mapbox/gm-navigation-app/pull/2205) cc @LukasPaczos @abhishek1508

Closing this and tracking other tickets and requests as mentioned in the comment above.

/cc @VitaminPSG
